src/protocols/displayport.stanza

```diff
@@ -0,0 +1,220 @@
+
+public pcb-bundle displayport :
+  name = "DisplayPort"
+  description = "DisplayPort Digital Display Interface"
+  port aux  : diff-pair
+  port lane : diff-pair[4]
+  port hpd  : pin ; hot plug detect
+
+
+defn constrain-ch (sw:Toleranced, ml:Double, a:JITXObject, b:JITXObject) :
+  skew-window(sw, a => b)
+  max-loss(ml, a => b)
+
+doc: \<DOC>
+@brief Construct a DisplayPort lane.
+
+This function will construct the lane connections of a DisplayPort connection. The `src` and `dst`
+arguments must both be of type `displayport`.
+
+@param sw The intra-pair skew time difference for the differential signals being connected.
+@param ml The maximum loss in dB for all differential signals being connected.
+@param cap Optional DC blocking capacitor to be inserted between the TX and RX pairs. Note that
+there are 2 blocking caps inserted, one on each of the P,N signals of the pair.
+@param src This is assumed to be the "Tx" side of the link. This is expected to be a port of `Bundle` type `displayport`
+@param dst This is assumed to be the "Rx" side of the link. This is expected to be a port of `Bundle` type `displayport`
+
+<DOC>
+public defn connect-displayport-lane (sw:Toleranced, ml:Double, cap:Instantiable|False, src:JITXObject, dst:JITXObject) :
+  inside pcb-module:
+    for i in indices(src.lane) do :
+      match(cap) :
+        (c:Instantiable) :
+          inst bl-cap : dp-coupler(c)
+          require tx1 : dual-pair from bl-cap
+          topo-net(src.lane[i], tx1.A)
+          topo-net(tx1.B, dst.lane[i])
+        (f:False) :
+          topo-net(src.lane[i], dst.lane[i])
+      constrain-ch(sw, ml, src.lane[i], dst.lane[i])
+
+doc: \<DOC>
+@brief Construct aux connection of a DisplayPort connection.
+
+This function will construct the AUX channel of a DisplayPort connection with the 
+optional blocking cap and pull-up resistors.
+The `src` and `dst` arguments must both be of type `displayport`.
+
+@param sw The intra-pair skew time difference for the differential signals being connected.
+@param ml The maximum loss in dB for all differential signals being connected.
+@param res Optional pull-up resistor to be added to the AUX pair before the capacitor. 
+Note that there are 2 resistors added, one on each of the P,N signals of the pair.
+@param vdd Power supply pin to be used for the optional pull-up resistors. Must be present if the resistors are used.
+@param cap Optional DC blocking capacitor to be inserted between the AUX pair. Note that
+there are 2 blocking caps inserted, one on each of the P,N signals of the pair.
+@param src This is assumed to be the source side of the link. This is expected to be a port of `Bundle` type `displayport`
+@param dst This is assumed to be the sink side of the link. This is expected to be a port of `Bundle` type `displayport`
+
+<DOC>
+public defn connect-displayport-aux (sw:Toleranced, ml:Double, res:Instantiable|False, vdd:JITXObject|False, cap:Instantiable|False, src:JITXObject, dst:JITXObject) :
```

Comment:
I think we should consider making a different polymorphic function here - or using named keyword arguments. 

This is going to get used as either: 

```
connect-displayport-aux(sw, ml, false, false, cap, src, dst) 
connect-displayport-aux(sw, ml, res, vdd, cap, src, dst) 

```

The `false` argument without any label or context is difficult to read. When I'm looking at this code without knowing the implementation - I will have no idea what the `false, false` is referring to. 

Having two variables that are dependent is also a non-optimal pattern. It is really easy to introduce bugs inadvertently. To wit - the case where the user passes `vdd/res` but not `cap` is not covered. 

There is also not much that is shared between these two cases that necessitates it all being in one function .

I would suggest creating: 

```
defn connect-displayport-aux (
  sw:Toleranced, ml:Double, 
  src:JITXObject, dst:JITXObject 
  -- 
  cap:Instantiable = ?,
  term:[Instantiable, JITXObject] = ?) :
 
```

The ` = ?` syntax forces a `Maybe<Instantiable>` or `Maybe<[Instantiable, JITXObject]>` 

You can then use the match with only two arguments which reduces the number of cases that you have to handle. 

It also forces the user to do: 

```
connect-displayport-aux(sw, ml,  src, dst, cap = bcap ) 
connect-displayport-aux(sw, ml, src, dst, cap = bcap, term = [r, vdd] ) 
```

So that now, there is no `false, false` and I know explicitly the arg names.

Reply:
The problem with `= ?` format is when you have to cascade it. 

You reference this function below in `connect-displayport` and using `= ?` in the parent and the child functions is not a great developer experience. You end up needing a `match` in the parent to deal with the different cases instead of deferring to the child.

Reply:
Because of that - the polymorphic approach might be a better option: 

```
defn connect-displayport-aux ( sw:Toleranced, ml:Double, src:JITXObject, dst:JITXObject, cap:Instantiable|False):
   ...

defn connect-displayport-aux ( sw:Toleranced, ml:Double, src:JITXObject, dst:JITXObject, cap:Instantiable|False, term:[Instantiable, JITXObject]|False) :

```
The second one can call the first one in the case where `term` is `False`.

Reply:
Thanks -- a lot nicer.